File: Atcoder_#317/javascript/c_remembering_the_days.js
function main(input) {
    const lines = input.trim().split('\n');
    const [n, m] = lines[0].split(' ').map(Number);
    const edges = Array.from({ length: n }, () => Array(n).fill(null));

    // 辺の情報を格納する
    for (let i = 1; i <= m; i++) {
        const [a, b, c] = lines[i].split(' ').map(Number);
        edges[a - 1][b - 1] = c;
        edges[b - 1][a - 1] = c;
    }

    // DPテーブルの初期化
    const dp = Array.from({ length: 1 << n }, () => Array(n).fill(-1));
    let maxDistance = 0;

    // 初期化: 各街を出発点に設定
    for (let i = 0; i < n; i++) {
        dp[1 << i][i] = 0;
    }

    // 全ての部分集合についてDPを計算
    for (let mask = 0; mask < (1 << n); mask++) {
        for (let u = 0; u < n; u++) {
            if (dp[mask][u] !== -1) {
                for (let v = 0; v < n; v++) {
                    if (edges[u][v] !== null && (mask & (1 << v)) === 0) {
                        const newMask = mask | (1 << v);
                        dp[newMask][v] = Math.max(dp[newMask][v], dp[mask][u] + edges[u][v]);
                        maxDistance = Math.max(maxDistance, dp[newMask][v]);
                    }
                }
            }
        }
    }

    // 結果を出力
    console.log(maxDistance);
}

// 標準入力を読み込み、main関数を呼び出し
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin', 'utf8');
main(input);

File: Atcoder_#317/ruby/c_remembering_the_days.rb
# 入力を受け取る
n, m = gets.split.map(&:to_i)
edges = Array.new(n) { Array.new(n, nil) }

# 辺の情報を格納する
m.times do
  a, b, c = gets.split.map(&:to_i)
  edges[a-1][b-1] = c
  edges[b-1][a-1] = c
end

# DPテーブルの初期化
dp = Array.new(1 << n) { Array.new(n, -1) }
max_distance = 0

# 初期化: 各街を出発点に設定
n.times do |i|
  dp[1 << i][i] = 0
end

# 全ての部分集合についてDPを計算
(1 << n).times do |mask|
  n.times do |u|
    if dp[mask][u] != -1
      n.times do |v|
        if edges[u][v] && (mask & (1 << v)) == 0
          new_mask = mask | (1 << v)
          dp[new_mask][v] = [dp[new_mask][v], dp[mask][u] + edges[u][v]].max
          max_distance = [max_distance, dp[new_mask][v]].max
        end
      end
    end
  end
end

# 結果を出力
puts max_distance
